### PR TITLE
ci: add typos check 

### PR DESCRIPTION
**Pull request by tisonkun**
This PR adds a typo checker CI action that will fail
future PRs if they introduce typos and spelling mistakes.

This supersedes and thus closes #1660.

### CODE REVIEW

File: src/commands/commander.h [hunk 1]
```diff
@@ -126,7 +126,7 @@ struct CommandAttributes {
   // negative number -n means number of arguments is equal to or large than n
   int arity;
 
-  // space-splitted flag strings to initialize flags
+  // space-sparated flag strings to initialize flags
```
Comment (torwig) [inline src/commands/commander.h]: I think there is another typo: `separated`.

Comment (tisonkun) [reply src/commands/commander.h]:
```suggestion
  // space-separated flag strings to initialize flags
```

🤣

Comment (tisonkun) [reply src/commands/commander.h]: Report to upstream dict - https://github.com/crate-ci/typos/issues/784#issuecomment-1673403677

**Review by torwig — DISMISSED** [review]
LGTM

File: .github/config/typos.toml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+[files]
+extend-exclude = [
+    "src/common/sha1.cc",
+    "tests/gocase/util/slot.go",
+]
```
Comment (szepeviktor) [inline .github/config/typos.toml]: Very nice to see Typos in action!

```suggestion
[files]
extend-exclude = [
    ".git/",
    "src/common/sha1.cc",
    "tests/gocase/util/slot.go",
]
ignore-hidden = false
```

This checks dotfiles too.
Typos works on the filesystem, not on git index.

**Comment by szepeviktor** [comment]
Currently "Install typos-cli" step takes 2-3 minutes 😲

This takes 1 second 🐎
https://github.com/szepeviktor/byte-level-care/blob/b2e8a838c14be862f95f1bb3486d3b2479b1bc79/.github/workflows/spelling.yml#L34-L46

**Comment by tisonkun** [comment]
> Currently "Install typos-cli" step takes 2-3 minutes 😲

I saw similar trick in OpenDAL. Let me try it out.

Actually we may make use of Rust GHA cache. But it's an action beyond the approved ones so I may postpone a bit for using it.

File: .github/config/typos.toml (hunk 2 above)
Comment (tisonkun) [reply .github/config/typos.toml]: Thank you!

Comment (szepeviktor) [reply .github/config/typos.toml]: Glad to contribute.

**Comment by tisonkun** [comment]
@szepeviktor I found that typos upstream provides an action already and thus I'm making use of it :D

Maybe you can switch to this solution also.

File: .github/workflows/kvrocks.yaml [hunk 3]
```diff
@@ -50,6 +50,16 @@ jobs:
...
 
+  check-typos:
+    name: Check typos
+    runs-on: ubuntu-22.04
+    env:
+      FORCE_COLOR: 1
+    steps:
+      - uses: actions/checkout@v3
+      - name: Run typos check
+        uses: crate-ci/typos@master
```
Comment (szepeviktor) [inline .github/workflows/kvrocks.yaml]: The configuration file can be specified
https://github.com/crate-ci/typos/blob/ef5fcf92dfbd679f97c0371159e143852f7b1eb1/action.yml#L28

```suggestion
      - name: Run typos check
        uses: crate-ci/typos@master
        with:
          config: .github/config/typos.toml
```

**Comment by szepeviktor** [comment]
> Maybe you can switch to this solution also.

@tisonkun Here are some facts for you.

- that action is very complicated - but you see only 2 lines in your workflow
- that action works on the filesystem, not on git index
- that action [runs `typos` twice](https://github.com/crate-ci/typos/blob/ef5fcf92dfbd679f97c0371159e143852f7b1eb1/action/entrypoint.sh#L64-L65) ❗

These made me rewrite it [in 6 commands](https://github.com/szepeviktor/byte-level-care/blob/b2e8a838c14be862f95f1bb3486d3b2479b1bc79/.github/workflows/spelling.yml#L38-L49).

File: .github/workflows/kvrocks.yaml (hunk 3 above)
Comment (tisonkun) [reply .github/workflows/kvrocks.yaml]: Thank you!

**Comment by tisonkun** [comment]
@szepeviktor thanks for your advice. Also the action isn't approved by INFRA yet. May you create a pr onto my with your installation solution? That is, open a PR against this one, Or use suggestions.

**Comment by szepeviktor** [comment]
I am sorry. I cannot contribute here.

My solution needs hard-wired excludes etc.

**Comment by PragmaTwice** [comment]
ASF repo does not allow to use action like `crate-ci/typos@master`. Maybe you need to specify the commit hash or version tag, like `@1234567` or `@1.0`, which is allowed.

**Comment by tisonkun** [comment]
@PragmaTwice No. It's the action itself not approved, not the version the issue.

So we can either ask INFRA to approve or use a non-action solution. I'm going to reuse @szepeviktor's solution in https://github.com/szepeviktor/byte-level-care under MIT license.

**Comment by tisonkun** [comment]
Or just OpenDAL's for simplicity.

**Comment by tisonkun** [comment]
@torwig @PragmaTwice Kvrocks failed to build. Do you have any idea - https://github.com/apache/kvrocks/actions/runs/5827809599/job/15805697484?pr=1663?

**Review by enjoy-binbin — DISMISSED** [review]

**Review by git-hulk — DISMISSED** [review]

File: .github/workflows/kvrocks.yaml [hunk 4]
```diff
@@ -134,12 +146,13 @@ jobs:
             with_sanitizer: -DENABLE_ASAN=ON
             without_jemalloc: -DDISABLE_JEMALLOC=ON
             compiler: clang
-          - name: Ubuntu GCC TSan
-            os: ubuntu-20.04
-            without_jemalloc: -DDISABLE_JEMALLOC=ON
-            with_sanitizer: -DENABLE_TSAN=ON
-            compiler: gcc
-            ignore_when_tsan: -tags="ignore_when_tsan"
+          # FIXME - tsan.so not found
```
Comment (tisonkun) [inline .github/workflows/kvrocks.yaml]: See also - https://github.com/apache/kvrocks/issues/1662

**Review by PragmaTwice — DISMISSED** [review]

**Review by enjoy-binbin — APPROVED** [review]

**Review by torwig — APPROVED** [review]

**Pull request merged**